### PR TITLE
📈 Add custom analytics event for bookend clicks

### LINKED ISSUE #23893: Track clicks in amp-story-bookend
Currently it isn't possible to track clicks coming from the `amp-story-bookend`. We should enable this behavior.

### PR DESCRIPTION
**Pull request by Enriqe**
Closes #23893

Adds a new analytics event trigger `story-bookend-click` and variables: 
* `storyBookendComponentPosition`
* `storyBookendComponentType`
* `storyBookendTargetHref`

### CODE REVIEW

File: extensions/amp-story/1.0/bookend/amp-story-bookend.js [hunk 1]
```diff
@@ -91,6 +93,14 @@ const REPLAY_ICON_TEMPLATE = {
 /** @const {string} */
 const TAG = 'amp-story-bookend';
 
+/** @enum {string} */
+const CLICKABLE_COMPONENT_TYPES = {
+  CTA_LINK: 'i-amphtml-story-bookend-cta-link',
+  LANDSCAPE: 'i-amphtml-story-bookend-landscape',
+  PORTRAIT: 'i-amphtml-story-bookend-portrait',
+  SMALL: 'i-amphtml-story-bookend-article',
+};
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/bookend/amp-story-bookend.js]: This is very fragile as the classeNames we actually set aren't from this ENUM. I think it's worth trying to see how we can change the way we set classNames/attributes to make it re-usable, instead of adding a new standalone ENUM.
As a bonus, depending on how you implement this, you might be able to get rid of the `getClickedComponentIndex_` and `getClickedComponentType_` methods.

**Comment by Enriqe** [comment]
Sounds good. Can you take a look at the new implementation while I write the e2e tests? Thanks!

File: extensions/amp-story/1.0/bookend/bookend-component.js [hunk 2]
```diff
@@ -160,13 +161,18 @@ export class BookendComponent {
...
+    const data = {position: 1};
     components.forEach(component => {
       const {type} = component;
       if (type && componentBuilderInstanceFor(type)) {
-        fragment.appendChild(
-          componentBuilderInstanceFor(type).buildElement(component, doc)
+        const el = componentBuilderInstanceFor(type).buildElement(
+          component,
+          doc,
+          data
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/bookend/bookend-component.js]: The position is unreliable here, since `prependTitle` could potentially add a component, offsetting everything by one.
We could replace this `data` with `{position: index}` where `index` comes from the `forEach`

File: extensions/amp-story/1.0/bookend/components/bookend-component-interface.js [hunk 3]
```diff
@@ -14,6 +14,30 @@
...
 
+/** @enum {string} */
+export const BOOKEND_COMPONENT_CLASSES = {
+  CTA_LINK: 'i-amphtml-story-bookend-cta-link',
+  LANDSCAPE: 'i-amphtml-story-bookend-landscape',
+  PORTRAIT: 'i-amphtml-story-bookend-portrait',
+  SMALL: 'i-amphtml-story-bookend-article',
+};
+
+export const BOOKEND_COMPONENT_TYPES = {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/bookend/components/bookend-component-interface.js]: Nit: jsdoc

File: extensions/amp-story/1.0/bookend/components/cta-link.js [hunk 4]
```diff
@@ -106,8 +112,15 @@ export class CtaLinkComponent {
       const refs = htmlRefs(el);
       refs['linkText'].textContent = currentLink['text'];
 
+      el[AMP_STORY_BOOKEND_COMPONENT_DATA] = {
+        position: data.position++,
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/bookend/components/cta-link.js]: Is this how it behaves?

- 0 heading
- 1 small
- 2 small
- 3 link
  - 4 link
  - 5 link
  - 6 link
- 4 small

File: extensions/amp-story/1.0/bookend/components/bookend-component-interface.js [hunk 5]
```diff
@@ -14,6 +14,30 @@
  * limitations under the License.
  */
 
+/** @enum {string} */
+export const BOOKEND_COMPONENT_CLASSES = {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/bookend/components/bookend-component-interface.js]: Why is this change needed?

File: extensions/amp-story/1.0/test/test-amp-story-bookend.js [hunk 6]
```diff
@@ -365,6 +367,99 @@ describes.realWin('amp-story-bookend', {amp: true}, env => {
...
+              'amphtml': true,
+            },
+          ],
+        },
+      ],
+    };
+
+    sandbox.stub(requestService, 'loadBookendConfig').resolves(userJson);
+    const analyticsSpy = sandbox.spy(bookend.analyticsService_, 'triggerEvent');
+    const variableSpy = sandbox.spy(
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/test/test-amp-story-bookend.js]: Can we let the variable service do its magic and assert on values from `variableService.get(...)`?

**Comment by estherkim** [comment]
@Enriqe for the e2e tests, it'll be helpful to wait for https://github.com/ampproject/amphtml/pull/24347 to merge. I don't know if `describes.integration` has mobile support but `describes.endtoend` will soon :)

File: extensions/amp-story/1.0/bookend/components/bookend-component-interface.js (hunk 5 above)
Comment (Enriqe) [reply extensions/amp-story/1.0/bookend/components/bookend-component-interface.js]: To fix a circular dependency between `/bookend/bookend-component.js` -> `/bookend/components/{article,portait,...}.js ` -> `/bookend/bookend-component.js`.

I could put them in `amp-story-bookend.js` too if you prefer?

File: extensions/amp-story/1.0/bookend/components/cta-link.js (hunk 4 above)
Comment (Enriqe) [reply extensions/amp-story/1.0/bookend/components/cta-link.js]: Nope, like this:

* 0 heading
* 1 small
* 2 small
* 3 link 
* 4 link
* 5 link 
* 6 small 

Hence the reduction of `data.position` below in line 123. I think this is what would be easier to understand for publishers. WDYT?

File: extensions/amp-story/1.0/bookend/components/bookend-component-interface.js (hunk 5 above)
Comment (gmajoulet) [reply extensions/amp-story/1.0/bookend/components/bookend-component-interface.js]: Sorry, my question was more around why we need this enum in the first place, over the existing implementation.

File: extensions/amp-story/1.0/bookend/components/cta-link.js (hunk 4 above)
Comment (Enriqe) [reply extensions/amp-story/1.0/bookend/components/cta-link.js]: SG. I will keep the position as the index of the component (links sharing the same position).

P.S. if you were curious about the decrementing one time: since the increment was happening in the inner loop of `cta-link.js`, then again in the outer loop of `bookend-component.js` I needed to decrement one time to avoid skipping a position. This should be less confusing with the new logic.

File: extensions/amp-story/1.0/bookend/components/bookend-component-interface.js (hunk 5 above)
Comment (Enriqe) [reply extensions/amp-story/1.0/bookend/components/bookend-component-interface.js]: Doh, sorry. This isn't needed anymore.

**Comment by Enriqe** [comment]
Cool @estherkim, thanks for the heads up! 

@gmajoulet this is ready for another review :)
